exercises/01.compartmentalisation/js.cc

```diff
@@ -0,0 +1,568 @@
+	                              mvm_Value *result,
+	                              mvm_Value *args,
+	                              uint8_t    argCount)
+	{
+		return call_export<ExportedFn<E>>(vm, result, args, argCount);
+	}
+
+	/**
+	 * Print a string passed from JavaScript.
+	 */
+	template<>
+	mvm_TeError exported_function<Print>(mvm_VM            *vm,
+	                                     mvm_HostFunctionID funcID,
+	                                     mvm_Value         *result,
+	                                     mvm_Value         *args,
+	                                     uint8_t            argCount)
+	{
+		// Iterate over the arguments.
+		for (unsigned i = 0; i < argCount; i++)
+		{
+			// Coerce the argument to a string and get it as a C string
+			const char *str = mvm_toStringUtf8(vm, args[i], nullptr);
+			// Write each character to the UART
+			for (; *str != '\0'; str++)
+			{
+				ImplicitUARTOutput::write(*str);
+			}
+		}
+		// Write a trailing newline
+		ImplicitUARTOutput::write('\n');
+		// Unconditionally return success
+		return MVM_E_SUCCESS;
+	}
+
+	/**
+	 * Callback from microvium that resolves imports.
+	 *
+	 * This resolves each function to the template instantiation of
+	 * `exported_function` with `funcID` as the template parameter.
+	 */
+	mvm_TeError
+	resolve_import(mvm_HostFunctionID funcID, void *, mvm_TfHostFunction *out)
+	{
+		return magic_enum::enum_switch(
+		  [&](auto val) {
+			  constexpr Exports Export = val;
+			  *out                     = exported_function<Export>;
+			  return MVM_E_SUCCESS;
+		  },
+		  Exports(funcID),
+		  MVM_E_UNRESOLVED_IMPORT);
+	}
+
+	/**
+	 * Helper that deletes a Microvium VM when used with a C++ unique pointer.
+	 */
+	struct MVMDeleter
+	{
+		void operator()(mvm_VM *mvm) const
+		{
+			mvm_free(mvm);
+		}
+	};
+} // namespace
+
+/// Thread entry point.
+void __cheri_compartment("js") run()
+{
+	// Set the secret value on startup.
+	set_secret();
+
+	while (true)
+	{
+		// Get a capability to the UART.
+		auto uart = MMIO_CAPABILITY(Uart, uart);
+		// Create a vector to hold the bytecode.
+		std::vector<uint8_t> bytecode;
+		// Helper that reads a character and errors if it isn't the expected
+		// one.
+		auto skip = [&](char byte) {
+			char c = uart->blocking_read();
+			Debug::Assert(byte == c, "Read '{}', expected '{}'", c, byte);
+		};
+		// Helper that reads from the UART until a specific character is seen
+		auto skipUntil = [&](char byte) {
+			while (uart->blocking_read() != byte) {}
+		};
+		// Helper that converts a hex character to an integer.  Returns 0 on
+		// invalid values.
+		auto hexByteToNumber = [](char c) {
+			switch (c)
+			{
+				case '0' ... '9':
+					return c - '0';
+				case 'a' ... 'f':
+					return c - 'a' + 10;
+				case 'A' ... 'F':
+					return c - 'A' + 10;
+			}
+			return 0;
+		};
+
+		// Read values in the form generated by microvium's hex output: a
+		// series of hex bytes in braces.
+		skipUntil('{');
```

Comment:
Maybe we should be warning if we discard something other than whitespace?

Reply:
If this were a real program, maybe, but it's already a bit overcomplicated since it's meant to be used only with the load script, which does the right thing.